参考自: https://www.zhihu.com/question/29683794/answer/251623934

# 一枚硬币，扔了10次都是正面朝上，再扔一次反面朝上的概率是多少？

假设：
H0：硬币是普通的，每一面朝上的概率都是50%。
H1：硬币是特殊的，导致某一面一直朝上。

## 频率学派
频率学派（Frequentist）的零假设显著性检验（NHST）,需要计算在H0成立的条件下，得到目前结果（data, D）的概率，记做p(D|H0)。这个p就是所谓的统计显著性，如果p值小于某个临界点，比如行为科学（就是心理学）研究里惯用的经验值0.05。

对于抛硬币问题，如果H0成立，那么每一次正反面朝上的概率都是50%，可以得到连续i次正面朝上的概率是0.5^i，这是一个单尾概率。

In [7]:
for(i in 1:20) {
    cat("i =",i,", p(D|H0) =",0.5^i,"\n")
}

i = 1 , p(D|H0) = 0.5 
i = 2 , p(D|H0) = 0.25 
i = 3 , p(D|H0) = 0.125 
i = 4 , p(D|H0) = 0.0625 
i = 5 , p(D|H0) = 0.03125 
i = 6 , p(D|H0) = 0.015625 
i = 7 , p(D|H0) = 0.0078125 
i = 8 , p(D|H0) = 0.00390625 
i = 9 , p(D|H0) = 0.001953125 
i = 10 , p(D|H0) = 0.0009765625 
i = 11 , p(D|H0) = 0.0004882812 
i = 12 , p(D|H0) = 0.0002441406 
i = 13 , p(D|H0) = 0.0001220703 
i = 14 , p(D|H0) = 6.103516e-05 
i = 15 , p(D|H0) = 3.051758e-05 
i = 16 , p(D|H0) = 1.525879e-05 
i = 17 , p(D|H0) = 7.629395e-06 
i = 18 , p(D|H0) = 3.814697e-06 
i = 19 , p(D|H0) = 1.907349e-06 
i = 20 , p(D|H0) = 9.536743e-07 


如果采用0.05的显著性标准，当连续5次正面朝上的时候，p = 0.03125 < 0.05，我们就有理由拒绝H0，认为这不是一枚普通硬币。当然，显著性水平还可以再严格一点，比如要求p < 0.005才行，那么当连续8次正面朝上的时候，我们就有理由拒绝H0。

## 贝叶斯学派的思路

![alt text](http://latex.codecogs.com/gif.latex?p%28H%7CD%29%20%3D%20%5Cfrac%7Bp%28D%7CH%29p%28H%29%7D%7Bp%28D%7CH%29p%28H%29%20&plus;%20p%28D%7C%5Coverline%7BH%7D%29p%28%5Coverline%7BH%7D%29%7D)

在贝叶斯公式里，分子中的p(D|H0)是我们之前求的p值，p(H0)是我们对H0成立与否的主观估计（先验概率），分母中的p(D|H1)是H1成立的条件下出现目前结果的概率，p(H1) = 1 - p(H0)。最后得到的p(H0|D)就是在已经得到目前数据的条件下，H0成立的后验概率。
这里我们需要事先估计两个概率：

- p(H0)：也就是主观估计这枚硬币是不是普通的。比如我觉得吧，世界上还是真币多一些，而且如果你说这枚硬币是普通的，我也比较相信你，所以我暂且估计p(H0) = 0.99。
- p(D|H1)：也就是对于一枚特殊硬币，出现这种连续正面朝上的概率是多少。

这种特制硬币，既可能被设计成一直正面朝上，也可能被设计成一直反面朝上，所以我暂且估计p(D|H1) = 0.5。当然，你也可以估计它等于1，相当于认为这种特殊硬币是专为正面朝上设计的


In [9]:
for(i in 1:20) {
  pH=0.99  # p(H0), 对普通硬币概率的主观估计
  p0=0.5^i  # p(D|H0), 如果是普通硬币(H0), 连续正面朝上的概率(单尾)
  p1=0.5  # p(D|H1), 如果是特殊硬币(H1), 连续正面朝上的概率(单尾)
  Bayes_P=p0*pH/(p0*pH+p1*(1-pH))
  cat("i =",i,", p(D|H0) =",p0,", Bayes p(H0|D) =",Bayes_P,"\n")
}


i = 1 , p(D|H0) = 0.5 , Bayes p(H0|D) = 0.99 
i = 2 , p(D|H0) = 0.25 , Bayes p(H0|D) = 0.980198 
i = 3 , p(D|H0) = 0.125 , Bayes p(H0|D) = 0.961165 
i = 4 , p(D|H0) = 0.0625 , Bayes p(H0|D) = 0.9252336 
i = 5 , p(D|H0) = 0.03125 , Bayes p(H0|D) = 0.8608696 
i = 6 , p(D|H0) = 0.015625 , Bayes p(H0|D) = 0.7557252 
i = 7 , p(D|H0) = 0.0078125 , Bayes p(H0|D) = 0.607362 
i = 8 , p(D|H0) = 0.00390625 , Bayes p(H0|D) = 0.4361233 
i = 9 , p(D|H0) = 0.001953125 , Bayes p(H0|D) = 0.2788732 
i = 10 , p(D|H0) = 0.0009765625 , Bayes p(H0|D) = 0.1620295 
i = 11 , p(D|H0) = 0.0004882812 , Bayes p(H0|D) = 0.08815672 
i = 12 , p(D|H0) = 0.0002441406 , Bayes p(H0|D) = 0.04611085 
i = 13 , p(D|H0) = 0.0001220703 , Bayes p(H0|D) = 0.02359952 
i = 14 , p(D|H0) = 6.103516e-05 , Bayes p(H0|D) = 0.01194066 
i = 15 , p(D|H0) = 3.051758e-05 , Bayes p(H0|D) = 0.006006188 
i = 16 , p(D|H0) = 1.525879e-05 , Bayes p(H0|D) = 0.00301214 
i = 17 , p(D|H0) = 7.629395e-06 , Bayes p(H0|D) = 0.001508342 
i = 18 , p(D|H0)

## 总结
借助统计学里最基础的假设检验思想，并结合频率学派和贝叶斯学派的不同思路，我们终于解决了这个“世纪难题”。总的来看，当出现连续10次左右正面朝上的情况，我们就有一定的理由去怀疑这枚硬币的真假，或者怀疑抛硬币的环境中有什么不寻常的地方。保守估计的话，连续16次正面朝上，基本就能肯定这枚硬币将会一直正面朝上了。
